### PR TITLE
fix(common): date pipe is giving error weeks

### LINKED ISSUE #33961: The Date pipe returns incorrect week numbers
Hi, I believe I've come across a bug in the date parsing. I haven't seen any indications that this is expected behavior, so I'm opening an issue here.

# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/common: ^6.1.9

### Is this a regression?

 Not that I've seen.

### Description

 
The Date pipe has some unexpected behavior when it comes to weeks. A week is defined as a period of exactly 7 days (according to Wikipedia), and according to ISO 8601, the first week of a year is the week containing the first Thursday of the year.

By these rules, 2019 has 52 weeks, and 2020 has 53 weeks.

E.g. the first week of 2020 is the week from 30th of December to the 5th of January. 
The Date pipe returns the period from December 29th to December 31st as week 53, and January 1st of 2020 as week 1. 

The date parsing has an extra week in 2019, which has 3 days, which seems like unexpected behavior.

## 🔬 Minimal Reproduction

 https://angular-issue-repro2-ycvycu.stackblitz.io

```js
{{ '2019-12-29T00:00:00' | date:'w' }}
```

### PR DESCRIPTION
**Pull request by ajitsinghkaler**
DDtae pipe is giving wrong week number made a change that will return correct week numbers for ifferent years

Fixes #33961

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/common/test/pipes/date_pipe_spec.ts [hunk 1]
```diff
@@ -71,6 +71,12 @@ import {JitReflector} from '@angular/platform-browser-dynamic/src/compiler_refle
     describe('transform', () => {
       it('should use "mediumDate" as the default format',
          () => expect(pipe.transform('2017-01-11T10:14:39+0000')).toEqual('Jan 11, 2017'));
+
+      it('should return first week if some dates fall in previous year but belong to next year accordint to ISO 8601 format', () => {
```
Comment (petebacondarwin) [inline packages/common/test/pipes/date_pipe_spec.ts:75]: Drive-by suggestion:

```suggestion
      it('should return first week if some dates fall in previous year but belong to next year according to ISO 8601 format', () => {
```

Comment (ajitsinghkaler) [reply packages/common/test/pipes/date_pipe_spec.ts]: @petebacondarwin thanks made the change

**Comment by ajitsinghkaler** [comment]
@AndrewKushnir flaky failures on ci

File: packages/common/src/i18n/format_date.ts [hunk 2]
```diff
@@ -382,8 +382,11 @@ function weekGetter(size: number, monthBased = false): DateFormatter {
       const today = date.getDate();
       result = 1 + Math.floor((today + nbDaysBefore1stDayOfMonth) / 7);
     } else {
-      const firstThurs = getFirstThursdayOfYear(date.getFullYear());
+      let firstThurs = getFirstThursdayOfYear(date.getFullYear());
       const thisThurs = getThursdayThisWeek(date);
+      if (thisThurs.getFullYear() !== firstThurs.getFullYear()) {
```
Comment (AndrewKushnir) [inline packages/common/src/i18n/format_date.ts:387]: Could you please add a comment here describing this logic?

File: packages/common/test/pipes/date_pipe_spec.ts [hunk 3]
```diff
@@ -71,6 +71,11 @@ import {JitReflector} from '@angular/platform-browser-dynamic/src/compiler_refle
     describe('transform', () => {
       it('should use "mediumDate" as the default format',
          () => expect(pipe.transform('2017-01-11T10:14:39+0000')).toEqual('Jan 11, 2017'));
+
+      it('should return first week if some dates fall in previous year but belong to next year according to ISO 8601 format',
+         () => {
+           expect(pipe.transform('2019-12-29T00:00:00', 'w')).toEqual('1');
```
Comment (AndrewKushnir) [inline packages/common/test/pipes/date_pipe_spec.ts:77]: I'd propose to extend this test to include a couple more days (2019-12-28 and 2019-12-30), so that it's clear when the switch from `52` to `1` actually happens. Also may be add the same checks for a leap year (2016).

**Review by AndrewKushnir — COMMENTED** [review]
Hi @ajitsinghkaler, thanks for the fix 👍 

I've left a couple comments, please have a look when you get a chance.

FYI I've added the "breaking change" label to this PR. Even though it's a bug fix, it changes the behavior that other developers may rely on.

I'd also like to ask @petebacondarwin to have a look at this change as I'm not super familiar with this code and having additional reviewer would be very helpful.

Thank you.

File: packages/common/src/i18n/format_date.ts [hunk 4]
```diff
@@ -382,8 +382,16 @@ function weekGetter(size: number, monthBased = false): DateFormatter {
...
+      let firstThurs = getFirstThursdayOfYear(date.getFullYear());
       const thisThurs = getThursdayThisWeek(date);
+      /*
+       If some days of a year are part of next year according
+       to ISO 8601. Change the firstThurs for those days to next
+       year firstThurs
+      */
+      if (thisThurs.getFullYear() !== firstThurs.getFullYear()) {
+        firstThurs = getFirstThursdayOfYear(date.getFullYear() + 1);
+      }
```
Comment (petebacondarwin) [inline packages/common/src/i18n/format_date.ts:394]: I believe we can make this logic much simpler...

```suggestion
      const thisThurs = getThursdayThisWeek(date);
      // Some days of a year are part of next year according to ISO 8601.
      // Compute the firstThurs from the year of this week's Thursday
      const firstThurs = getFirstThursdayOfYear(thisThurs.getFullYear());
```

WDYT? Am I missing something?

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Thanks for this fix. It appears to be correct. 
I suggest a refactoring that would be less code and a bit simpler.

File: packages/common/src/i18n/format_date.ts (hunk 4 above)
Comment (ajitsinghkaler) [reply packages/common/src/i18n/format_date.ts:394]: seems much better to me

File: packages/common/test/pipes/date_pipe_spec.ts [hunk 5]
```diff
@@ -71,6 +71,20 @@ import {JitReflector} from '@angular/platform-browser-dynamic/src/compiler_refle
...
+         () => {
+           expect(pipe.transform('2019-12-28T00:00:00', 'w')).toEqual('52');
+           expect(pipe.transform('2019-12-29T00:00:00', 'w')).toEqual('1');
+           expect(pipe.transform('2019-12-30T00:00:00', 'w')).toEqual('1');
+         });
+
+      it('should return first week if some dates fall in previous leap year but belong to next year according to ISO 8601 format',
+         () => {
+           expect(pipe.transform('2012-12-29T00:00:00', 'w')).toEqual('52');
+           expect(pipe.transform('2019-12-30T00:00:00', 'w')).toEqual('1');
```
Comment (AndrewKushnir) [inline packages/common/test/pipes/date_pipe_spec.ts:85]: @ajitsinghkaler thanks for adding more test cases!

It looks like the year is wrong in this case and in the one below (2019 instead of 2012).

**Review by petebacondarwin — APPROVED** [review]
LGTM

**Comment by AndrewKushnir** [comment]
FYI I've started [presubmit](http://test/OCL:317675331:BASE:317675367:1592845083327:b67ffbd0) in Google's codebase, will let you know the result as soon as I have it. Thank you.

**Review by AndrewKushnir — APPROVED** [review]
Thanks for the fix @ajitsinghkaler 👍

**Comment by AndrewKushnir** [comment]
Hi @ajitsinghkaler, just want to give a quick update. Running tests in Google's codebase didn't indicate any issues, which is great. Also given the fact that this is a bug fix with low risk, I'm removing the "breaking changes" flag. In any case, since the behavior of the framework changes slightly (to become more accurate), we would merge it to master only and it will be released as a part of the next minor version (v10.1.0). Thank you.
